File: emu/emu_plot/python/plot_samp.py
# Read Sampling Tool output

import os
import numpy as np
import glob
import matplotlib.pyplot as plt
import global_emu_var as emu

def plot_samp(frun):
    # Define the byte order ('>' for big-endian)
    byte_order = '>'

    # Set EMU output file directory
    ff = os.path.join(frun, 'output')
#    print(f"ff is {ff}") # debugging 

    # Read samp.out_? (OBJF time-series)
    fdum = 'samp.out_'
    fdum_all = fdum + '*'
    aa = glob.glob(os.path.join(ff, fdum_all))
    aa.sort()

#    print(f"Files matching {fdum_all}: {aa}")  # Debugging print

    if len(aa) != 1:
        print()
        if len(aa) == 0:
            print('*********************************************')
            print(f'File {fdum_all} not found ... ')
            print()
        else:
            print('*********************************************')
            print(f'More than one {fdum_all} found ... ')
            print()
        return

    ip = aa[0].find(fdum) + len(fdum)
    frec = aa[0][ip:]
    nrec = int(frec)

    smp = np.zeros(nrec, dtype=np.float32)
    smp_mn = 1.0

    with open(aa[0], 'rb') as f:
        smp = np.fromfile(f, dtype=byte_order+'f4', count=nrec)
        smp_mn = np.fromfile(f, dtype=byte_order+'f4', count=1)[0]

    emu.smp=smp
    emu.smp_mn=smp_mn

    print()
    print('*********************************************')
    print('Read variables')
    print('   smp: temporal anomaly of sampled variable')
    print('   smp_mn: reference time-mean of sampled variable')
    print(f'from file {aa[0]}')

#    # Print the first five elements of smp
#    print(f"The first five elements of smp are: {smp[:5]}")


    # Read samp.step_? (time-step)
    fdum = 'samp.step_' + frec
    aa = glob.glob(os.path.join(ff, fdum))
    aa.sort()

    if len(aa) == 0:
        print()
        print('*********************************************')
        print(f'File {fdum} not found ... ')
        print()
        return

    smp_hr = np.zeros(nrec, dtype=np.float32)

    with open(aa[0], 'rb') as f:
        smp_hr = np.fromfile(f, dtype=byte_order+'i4', count=nrec)

    emu.smp_hr=smp_hr

    print()
    print('*********************************************')
    print('Read variable')
    print('   smp_hr: sample time (hours from 1/1/1992 12Z)')
    print(f'from file {aa[0]}')

    smp_yday = (smp_hr /24.) / 365.0 + 1992.0

    # Plot
    print()
    print('Plotting sampled time-series ... ')

    samp_t = smp_yday
    samp_v = smp + smp_mn

    frun_file = os.path.basename(frun)

    tmin = int(np.floor(np.min(samp_t))) - 1
    tmax = int(np.ceil(np.max(samp_t))) + 1

    plt.figure()
    plt.plot(samp_t, samp_v, label='smp + smp_mn')
    plt.title(frun_file)
    plt.xlabel('smp_hr')
    plt.ylabel('smp + smp_mn')
    plt.xlim(tmin, tmax)
    plt.autoscale(axis='y')
    plt.grid(True)
    plt.legend()

    plt.ion()  # Enable interactive mode
    plt.show(block=False)  # Show the plot without blocking

File: emu/emu_plot/matlab/plot_samp.m
function plot_samp(frun)
    % PLOT_SAMP  Read and plot EMU sampling tool output
    %
    %   Input:
    %     frun - path to EMU run directory (e.g., 'emu_samp_m_2_45_585_1')

    global emu
    byte_order = 'ieee-be';  % Big-endian
    ff = fullfile(frun, 'output');

    % ----------------------------------------
    % Read samp.out_? (temporal anomaly + mean)
    % ----------------------------------------
    files = dir(fullfile(ff, 'samp.out_*'));
    if numel(files) ~= 1
        if isempty(files)
            fprintf('*********************************************\n');
            fprintf('File samp.out_* not found ...\n\n');
        else
            fprintf('*********************************************\n');
            fprintf('More than one samp.out_* file found ...\n\n');
        end
        return
    end

    fname = fullfile(ff, files(1).name);
    rec_tag = extractAfter(files(1).name, 'samp.out_');
    nrec = str2double(rec_tag);

    fid = fopen(fname, 'r', byte_order);
    smp = fread(fid, nrec, 'float32');
    smp_mn = fread(fid, 1, 'float32');
    fclose(fid);

    emu.smp = smp;
    emu.smp_mn = smp_mn;

    fprintf('\n*********************************************\n');
    fprintf('Read variables\n');
    fprintf('   smp: temporal anomaly of sampled variable\n');
    fprintf('   smp_mn: reference time-mean of sampled variable\n');
    fprintf('from file %s\n', fname);

    % ----------------------------------------
    % Read samp.step_? (sample times)
    % ----------------------------------------
    fname_step = fullfile(ff, ['samp.step_' rec_tag]);
    if ~isfile(fname_step)
        fprintf('\n*********************************************\n');
        fprintf('File %s not found ...\n\n', ['samp.step_' rec_tag]);
        return
    end

    fid = fopen(fname_step, 'r', byte_order);
    smp_hr = fread(fid, nrec, 'int32');
    fclose(fid);

    emu.smp_hr = smp_hr;

    fprintf('\n*********************************************\n');
    fprintf('Read variable\n');
    fprintf('   smp_hr: sample time (hours from 1/1/1992 12Z)\n');
    fprintf('from file %s\n', fname_step);

    % ----------------------------------------
    % Time conversion and plot
    % ----------------------------------------
    smp_yday = smp_hr / 24 / 365 + 1992;  % Convert hours to decimal year
    samp_t = smp_yday;
    samp_v = smp + smp_mn;

    [~, frun_name] = fileparts(frun);
    tmin = floor(min(samp_t)) - 1;
    tmax = ceil(max(samp_t)) + 1;

    fprintf('\nPlotting sampled time-series ...\n');

    figure;
    plot(samp_t, samp_v, 'DisplayName', 'smp + smp\_mn');
    title(frun_name, 'Interpreter', 'none');
    xlabel('Time (decimal year)');
    ylabel('Sampled value');
    xlim([tmin tmax]);
    grid on;
    legend show;
end
